vim-engine/src/main/kotlin/com/maddyhome/idea/vim/common/Graphemes.kt

```diff
@@ -0,0 +1,255 @@
+package com.maddyhome.idea.vim.common
+
+/**
+ * Move over unicode extended grapheme cluster boundaries.
+ *
+ * https://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries
+ */
+public object Graphemes {
+  public fun next(charSeq: CharSequence, start: Int): Int? {
+    if (start < 0 || start >= charSeq.length) return null
+
+    return charSeq.nextBoundary(
+      start,
+      next = Int::plus,
+      nextCode = { if (it < length) Character.codePointAt(this, it) else null },
+      prevCode = { current, _ -> if (current > 0) Character.codePointBefore(this, current) else null },
+      swap = false,
+    )
+  }
+
+  public fun prev(charSeq: CharSequence, start: Int): Int? {
+    if (start <= 0 || start > charSeq.length) return null
+
+    return charSeq.nextBoundary(
+      start,
+      next = Int::minus,
+      nextCode = { if (it > 0) Character.codePointBefore(this, it) else null },
+      prevCode = { current, charCount ->
+        if (current - charCount > 0)
+          Character.codePointBefore(this, current - charCount)
+        else null
+      },
+      swap = true,
+    )
+  }
+}
+
+private inline fun CharSequence.nextBoundary(
+  start: Int,
+  crossinline next: (Int, Int) -> Int,
+  crossinline nextCode: CharSequence.(Int) -> Int?,
+  crossinline prevCode: CharSequence.(Int, Int) -> Int?,
+  swap: Boolean,
+): Int {
+  var current = start
```

Comment:
I see a kind of a fancy logic here. Don't you think there is a library for that by java or unicode?

Reply:
There is the library by the unicode organisation to handle unicode called ICU4J. But it is a pretty big library and we only need a tiny fraction of its functionality, I got a sense that idea-vim avoids unnecessary dependencies, so decided to implement it myself, given that the ruleset is fairly straightforward and compact (around a dozen rules total). Also, their API isn't great for this, the way they expose the functionality is through the implementation of the `java.text.BreakIterator` interface, which is a little awkward to use for this. 

We could technically switch to ICU4J, but I'm not sure of the merits of this switch.

*from the top-level comment*
> Also, you mentioned that the code is tested against test cases provided by unicode. Was this a manual testing? Is there is chance to inlude these test cases to the IdeaVim tests?

Yes, the technical report that introduces the notion of the extended grapheme cluster has a file with the test cases. I'll add it to the repo and write some tests that make use of it.

Reply:
Regarding ICU4J: let's keep your implementation then, but please add a note with the rationale about implementing this code by hand.